### PR TITLE
Rewrite filename trunction to handle UTF8 correctly, add test

### PR DESCRIPTION
**Pull request by bengotow**
Based on Slack conversation here: https://nylas.slack.com/archives/sync/p1470352819000868, and observation that our filename column can actually support 255 4-byte UTF8 strings just fine (https://dev.mysql.com/doc/refman/5.6/en/charset-unicode-utf8mb4.html). Seems we've been truncating filenames unnecessarily for a while.

Sidenote: I'm totally new to Python 2's handling of unicode, so this might be totally borked, but the test seems to work...

### CODE REVIEW

File: inbox/models/message.py [hunk 1]
```diff
@@ -30,12 +31,21 @@
...
+
+    # If `s` is not stored as a unicode string, but contains unicode
+    # characters, len will return the wrong value (bytes not chars).
+    # Convert it to unicode first.
+    if not isinstance(s, unicode):
+        s = s.decode('utf-8', 'ignore')
+
+    if len(s) > max_len:
+        # If we need to truncate the string, keep the extension
+        filename, fileext = os.path.splitext(s)
```
Comment (mhahnenberg) [inline inbox/models/message.py]: Dunno if it matters, but what happens when the extension is reeeeallly long?

Comment (bengotow) [inline inbox/models/message.py]: Good call - added some more test cases to account for that

**Comment by spang** [comment]
Can you include a test case that passes in a filename that's already a unicode object? Otherwise, LGTM!

The important things here, I'd say, are to make sure we're never truncating bytestrings. Really we should be decoding text to unicode objects as soon as we can, always pass unicode objects around for internal processing, and then only reencode to UTF-8 when needed for sending across the wire or serializing to the database. (I think SQLAlchemy correctly deals with the latter, though.)

**Comment by bengotow** [comment]
Sounds good! I think https://github.com/nylas/sync-engine/pull/357/files#diff-5b0543c68e6db7d52885d28d3cde74efR7 tests the unicode case. 👍

**Comment by spang** [comment]
Hmm @bengotow I'm not sure that's true! In those test cases, cname is a unicode bytestring, not a unicode _object_, and it won't evaluate as True for your `isinstance` check:

```
>>> cname = '\xf0\x9f\x87\xba\xf0\x9f\x87\xb8\xe2\x99\xb8\xf0\x9f\x98\x82.txt'
>>> type(cname)
<type 'str'>
>>> cname.decode('utf-8')
u'\U0001f1fa\U0001f1f8\u2678\U0001f602.txt'
>>> type(cname.decode('utf-8'))
<type 'unicode'>
>>> isinstance(cname, unicode)
False
>>> isinstance(cname.decode('utf-8'), unicode)
True
```
